pkg/kubectl/cmd/edit.go

```diff
@@ -0,0 +1,419 @@
+				fmt.Fprintln(cmd.Out(), results.AddError(err, info))
+				return nil
+			}
+			info.Refresh(updated, true)
+			fmt.Fprintf(out, "%s/%s\n", info.Mapping.Resource, info.Name)
+			return nil
+		})
+		if err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+
+		if results.retryable > 0 {
+			fmt.Fprintf(cmd.Out(), "You can run `kubectl update -f %s` to try this update again.\n", file)
+			return errExit
+		}
+		if results.conflict > 0 {
+			fmt.Fprintf(cmd.Out(), "You must update your local resource version and run `kubectl update -f %s` to overwrite the remote changes.\n", file)
+			return errExit
+		}
+		if len(results.edit) == 0 {
+			if results.notfound == 0 {
+				os.Remove(file)
+			} else {
+				fmt.Fprintf(cmd.Out(), "The edits you made on deleted resources have been saved to %q\n", file)
+			}
+			return nil
+		}
+
+		// loop again and edit the remaining items
+		infos = results.edit
+	}
+	return nil
+}
+
+// editReason preserves a message about the reason this file must be edited again
+type editReason struct {
+	head  string
+	other []string
+}
+
+// editHeader includes a list of reasons the edit must be retried
+type editHeader struct {
+	reasons []editReason
+}
+
+// WriteTo outputs the current header information into a stream
+func (h *editHeader) WriteTo(w io.Writer) error {
+	fmt.Fprint(w, `# Please edit the object below. Lines beginning with a '#' will be ignored,
+# and an empty file will abort the edit. If an error occurs while saving this file will be
+# reopened with the relevant failures.
+#
+`)
+	for _, r := range h.reasons {
+		if len(r.other) > 0 {
+			fmt.Fprintf(w, "# %s:\n", r.head)
+		} else {
+			fmt.Fprintf(w, "# %s\n", r.head)
+		}
+		for _, o := range r.other {
+			fmt.Fprintf(w, "# * %s\n", o)
+		}
+		fmt.Fprintln(w, "#")
+	}
+	return nil
+}
+
+// editResults capture the result of an update
+type editResults struct {
+	header    editHeader
+	retryable int
+	notfound  int
+	conflict  int
+	edit      []*resource.Info
+	file      string
+
+	delta   *jsonmerge.Delta
+	version string
+}
+
+func (r *editResults) AddError(err error, info *resource.Info) string {
+	switch {
+	case errors.IsInvalid(err):
+		r.edit = append(r.edit, info)
+		reason := editReason{
+			head: fmt.Sprintf("%s %s was not valid", info.Mapping.Kind, info.Name),
+		}
+		if err, ok := err.(kclient.APIStatus); ok {
+			if details := err.Status().Details; details != nil {
+				for _, cause := range details.Causes {
+					reason.other = append(reason.other, cause.Message)
+				}
+			}
+		}
+		r.header.reasons = append(r.header.reasons, reason)
+		return fmt.Sprintf("Error: the %s %s is invalid", info.Mapping.Kind, info.Name)
+	case errors.IsNotFound(err):
+		r.notfound++
+		return fmt.Sprintf("Error: the %s %s has been deleted on the server", info.Mapping.Kind, info.Name)
+
+	case errors.IsConflict(err):
+		if r.delta != nil {
+			v1 := info.ResourceVersion
+			if perr := applyPatch(r.delta, info, r.version); perr != nil {
```

Comment:
@deads2k, this applyPatch() approach doesn't work for a list of objects, because `r.delta` is for the list and `info` is for individual object.